### PR TITLE
[RFC] Add an API to remove autograd hooks from DDP

### PR DESCRIPTION
**Pull request by pritamdamania87**
Summary:
When creating a new DDP instance for the same model when an old DDP instance existed, the autograd hooks from the old DDP instance might not be cleared. Also, relying on python gc to clear out old autograd hooks is fragile and may not work 100% of the time.

As a result, in this PR I'm adding a way to explicitly remove these hooks from DDP

Test Plan:
Unit test added

cc @mrshenli @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### CODE REVIEW

**Comment by pritamdamania87** [comment]
@mrshenli @zhaojuanmao @rohan-varma Curious if you had any thoughts on whether this proposal makes sense :)

**Review by zhaojuanmao — APPROVED** [review]
looks good, thank you @pritamdamania87 :)

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -9411,6 +9411,59 @@ def buffer_comm_hook(ddp, named_buffers):
...
+                def forward(self, inp):
+                    if self.error:
+                        return self.fc1(SimulateError.apply(inp))
+                    else:
+                        return self.fc1(inp)
+
+
+            # Run with error to trigger backward pass that marks fc1 as being marked
+            # ready. If we don't remove autograd hooks before running below it would
+            # fail on the old autograd hook.
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:9445]: what would the error be? The `SimulateError`, or marked as ready twice issue?

**Review by rohan-varma — APPROVED** [review]
LGTM

File: torch/testing/_internal/distributed/distributed_test.py (hunk 1 above)
Comment (pritamdamania87) [reply torch/testing/_internal/distributed/distributed_test.py:9445]: The error would be the parameter is marked ready as twice.

**Pull request closed without merging**